uamqp/management_link.py

```diff
@@ -4,92 +4,185 @@
-        self._on_mgmt_error = kwargs.get('on_mgmt_error')
+        self._request_link = session.create_sender_link(  # type: SenderLink
+            endpoint,
+            on_link_state_change=self._on_sender_state_change,
+            send_settle_mode=SenderSettleMode.Unsettled,
+            rcv_settle_mode=ReceiverSettleMode.First
+        )
+        self._response_link = session.create_receiver_link(  # type: ReceiverLink
+            endpoint,
+            on_link_state_change=self._on_receiver_state_change,
+            on_message_received=self._on_message_received,
+            send_settle_mode=SenderSettleMode.Unsettled,
+            rcv_settle_mode=ReceiverSettleMode.First
+        )
+        self._on_amqp_management_error = kwargs.get('on_amqp_management_error')
+        self._on_amqp_management_open_complete = kwargs.get('on_amqp_management_open_complete')
+
+        self._status_code_field = status_code_field
+        self._status_description_field = status_description_field
+
+        self._sender_connected = False
+        self._receiver_connected = False
 
     def __enter__(self):
         self.open()
         return self
     
     def __exit__(self, *args):
         self.close()
-    
-    def _set_state(self, new_state):
-        previous_state = self.state
-        self.state = new_state
-        if new_state == ManagementLinkState.ERROR and self._on_mgmt_error:
-            self._on_mgmt_error()
 
     def _on_sender_state_change(self, previous_state, new_state):
+        _LOGGER.info("Management link sender state changed: %r -> %r", previous_state, new_state)
         if new_state == previous_state:
             return
-        #if self.state == ManagementLinkState.OPENING:
-        #    if new_state == LinkState.OPENING:
-        #elif self.state == ManagementLinkState.OPEN:
+        if self.state == ManagementLinkState.OPENING:
+            if new_state == LinkState.ATTACHED:
+                self._sender_connected = True
+                if self._receiver_connected:
+                    self.state = ManagementLinkState.OPEN
+                    self._on_amqp_management_open_complete(ManagementOpenResult.OK)
+            elif new_state in [LinkState.DETACHED, LinkState.DETACH_SENT, LinkState.DETACH_RCVD, LinkState.ERROR]:
+                self.state = ManagementLinkState.IDLE
+                self._on_amqp_management_open_complete(ManagementOpenResult.ERROR)
+        elif self.state == ManagementLinkState.OPEN:
+            if new_state is not LinkState.ATTACHED:
+                self.state = ManagementLinkState.ERROR
+                self._on_amqp_management_error()
         elif self.state == ManagementLinkState.CLOSING:
-            if new_state not in [LinkState.DETACHED, LinkState.DETACH_RCVD]:
-                self._set_state(ManagementLinkState.ERROR)
+            if new_state not in [LinkState.DETACHED, LinkState.DETACH_SENT, LinkState.DETACH_RCVD]:
```

Comment:
DETACH_SENT/DETACH_RCVD don't actually work yet..... The C implementation only had one state to represent these two - which is how I started - but then though I might introduce this as two separate states. No need to change this now - just a headsup